proxy/coco/src/peer/announcement.rs

```diff
@@ -0,0 +1,283 @@
+//! Compute, track and announce noteworthy changes to the network.
+
+use std::collections::HashSet;
+use std::ops::Deref as _;
+
+use kv::Codec as _;
+
+use librad::net::peer::{Gossip, Rev};
+use librad::uri::path::ParseError;
+use librad::uri::{Path, RadUrn};
+
+use crate::oid::Oid;
+use crate::state::Lock;
+
+/// Name for the bucket used in [`kv::Store`].
+const BUCKET_NAME: &str = "announcements";
+/// Key for the single value used as cache.
+const KEY_NAME: &str = "latest";
+
+/// Announcement errors.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Stop-gap until we get rid of crate level errors.
+    // TODO(xla): Remove once we transitioned to per module errors.
+    #[error(transparent)]
+    Crate(#[from] crate::error::Error),
+
+    /// Failures from [`kv`].
+    #[error(transparent)]
+    Kv(#[from] kv::Error),
+
+    /// Failures parsing.
+    #[error(transparent)]
+    Parse(#[from] ParseError),
+}
+
+/// An update and all the required information that can be announced on the network.
+pub type Announcement = (RadUrn, Oid);
+
+/// Unique list of [`Announcement`]s.
+pub type Updates = HashSet<Announcement>;
+
+/// Announces the list of given `updates` with the [`librad::net::protocol`].
+///
+/// # Errors
+///
+/// * if the announcemnet of one of the project heads failed
+pub async fn announce(state: Lock, updates: impl Iterator<Item = &Announcement> + Send) {
+    for (urn, hash) in updates {
+        let state = state.lock().await;
+        let protocol = state.api.protocol();
+
+        let have = Gossip {
+            urn: urn.clone(),
+            rev: Some(Rev::Git((*hash).into())),
+            origin: None,
+        };
+        protocol.announce(have).await;
+    }
+}
+
+/// Builds the latest list of [`Announcement`]s for the current state of the peer.
+///
+/// # Errors
+///
+/// * if listing of the projects fails
+/// * if listing of the Refs for a project fails
+pub async fn build(state: Lock) -> Result<Updates, Error> {
+    let state = state.lock().await;
+    let mut list: Updates = HashSet::new();
+
+    match state.list_projects() {
+        // TODO(xla): We need to avoid the case where there is no owner yet for the peer api, there
```

Comment:
Is there an issue for this, I have an idea :ghost:

Reply:
@FintanH You reckon it's a link issue, or something we should track in upstream?

Reply:
Hrmmm good question. My idea is that we could use the ol' `PhantomData` trick to make sure we get a `Store` that has its default rad self is set. I'll create an issue in link and can discuss there.

Reply:
https://github.com/radicle-dev/radicle-link/issues/316

Reply:
Looks like another thing to look at in #890